### DISCUSSION TITLE
Учёт записей в одной таблице из разных таблиц.

### DISCUSSION THREAD

**Discussion by Ekhorikov2** [totumonline/totum-ru-issues-and-discussions#253] [Как это сделать?] [answered]
Всем привет.
Возник у меня интересный вопрос:

**Теория**
Есть две таблицы циклов **1 Заказы клиентов** и **2 Заказы поставщикам**
В каждой из них, соответственно, хранятся расчётные таблицы (обзовём их документами, ЗАКАЗАМИ) Один документ заказывает товар у поставщика, а второй документ этот товар продаёт клиенту. 
В каждом из этих документов есть одна общая особенность - это **платежи**. **Платежи** это отдельная **простая таблица**. 
Из каждого из заказов, настроена кнопка, которая по **linkToTable** открывает таблицу **Платежи** и передаёт idшник в префильтр и таким образом позволяет привязать платежи из общей таблицы **Платежи** к каждому отдельному **Заказу клиента** или **Заказу поставщика**.

**Практика**
В теории всё красиво, а вот на практике.... получаем проблему одинаковых id в двух источниках циклов. Т.е. id **Заказа клиента** будет совпадать с id **Заказа поставщику** и в результате это приводит к тому, что например **Платежи** по **Заказу клиента с id15** будут отображаться в **Заказе поставщику с id15**!
По логике, необходимо задать оригинальные признаки **Платежам** из одного документа и из другого, в идеале, наверно это нужно сделать в id, т.к. во всех иных случаях, эти признаки префильтр не увидит, чтоб отличить одно от другого. А признаки внешние, можно вывести в строчную часть таблицы.

В общем вопрос №1 у кого есть идеи как такую штуку реализовать, может быть у кого-то есть другое видение реализации
Вопрос №2 какие варианты решения именно моего видения решения с, например, префиксами в id ( ЗК00012 и ЗП00012) хотя этот вариант наверно мало вероятен, т.к. префильтр префиксы не поймёт? В общем как задать оригинальный признак в зависимости от источника?
У меня были идей с id, но выше написал что не рабочая схема (или я что-то упустил). Была идея создать дополнительное поле, в котором задаётся дополнительный отличительный параметр, но в таком случае в префильтр этот, второй, параметр не передать, в общем интересно послушать ваше мнение!
Спасибо.
![image](https://user-images.githubusercontent.com/112487577/198314720-7124119e-29db-48bf-b93b-83ba5c498f21.png)

**Comment by erpic-ru** [comment]
В чем проблема добавить тип документа и дополнительно фильтровать по нему? Хоть десять параметров можно передать. и на другие документы расширить эту модель

**Reply by ttm-alexeypolunin** [reply]
Есть в описанном примере некоторая логика. Но на поверхности использование `Тип документа` для распределения — `исходящие` к одному, `входящие` к другому.

**Reply by erpic-ru** [reply]
Там ещё могут быть всякие фокусы с возвратами, поэтому надежней будет либо добавить поле "Тип связанного документа", либо генерировать уникальный номер и фильтровать по нему, в целом, тоже норм вариант

**Reply by ttm-alexeypolunin** [reply]
Да, вот я не мог понять что надо сделать с уникальным номером — я бы передавал номер с префиксом. В таблице с платежами сделал бы еще одно поле в котором на основе префикса вычислял бы к какой таблице циклов этот платеж относится.

Например:

— `p-15`
— `s-15`

И в строчной части в поле вычисляющем привязку:

```
=: if(condition: $split[0] = "p"; then: ...; else: ...)

split: strSplit(str: #number; separator: "-")
```

**Reply by Ekhorikov2** [reply]
> В чем проблема добавить тип документа и дополнительно фильтровать по нему? Хоть десять параметров можно передать. и на другие документы расширить эту модель

Вариант с несколькими префильтрами полем  **Тип документа** Кажется наиболее простым.
Добавил два скрытых технических поля id и Тип документа, первое  для работы ссылки из поля **Связанные документы** и склеивания корректного нейма со вторым (чтоб начиналось с Проект или Заказ поставщику) а второе чтоб фильтровать по типу вместе с id.

![image](https://user-images.githubusercontent.com/112487577/198349965-f52b6550-0c33-4e68-a9eb-d87008518ef1.png)

Вот только у меня вопрос возник  -  при создании платеже из Проектов (Заказов клиентов), **Тип документа** (Проект) передаётся при создании платежа, через

= : if (condition: #nomer_proekta !=; then: $list) 
                    list: select (table: 'zakaz'; cycle: #nomer_proekta ;  field: 'h_tip_dokumenta')

И ссылается поле именно на таблицу Проект (Заказ), а вот как поставить условие брать из таблицы **Заказы поставщику**, сообразить не могу? Или это нужно как-то в обратную сторону делать? Т.е. как передать в таблицу значение которое говорит о Типе документа?

**Reply by ttm-alexeypolunin** [reply]
Завтра сгенерирую примерчик на эту тему

**Reply by Ekhorikov2** [reply]
> Завтра сгенерирую примерчик на эту тему

Круто, спасибо большое.
И ещё вопрос по поводу idшников с префиксами, тоже интересная тема. Можно в двух словах, как их правильно сформировать? Вряд ли же это склеиванием полей делается?))

**Reply by Ekhorikov2** [reply]
> Завтра сгенерирую примерчик на эту тему

Я вот было подумал что можно передать значение поля через linkToTable, при нажатии кнопки, только такое заполнение как-то странно работает - таблица открывается, нажимаем **Добавить**, и поле остаётся пустым, закрываем строку, нажимаем ещё раз **Добавить** поле заполняется нужным значением.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Вот у нас простая таблица. В ней два поля Префильтра — один номер цикла, второй тип цикла:

<img width="555" alt="Снимок экрана 2022-10-28 в 18 02 58" src="https://user-images.githubusercontent.com/55755565/198667138-2102924c-a30e-414a-9e22-1e9be5d329d8.png">

Когда мы открываем ее из цикла мы передаем два префильтра и видим только то, что относится к нужному типу и номеру:

<img width="863" alt="Снимок экрана 2022-10-28 в 18 03 23" src="https://user-images.githubusercontent.com/55755565/198667621-6a591ea5-3700-43ac-aaba-324242e542c3.png">

Вот код в кнопке:

```
= : linkToTable(table: 'ttm16_simple'; title: "some title"; filter: 'fl_connect_with_cycle' = $#nci; filter: 'fl_cycle_type' = 1; target: "iframe"; width: 800)
```

+ `fl_connect_with_cycle` — передаем номер цикла.

+ `fl_cycle_type` — передаем тип цикла (в этом примере у меня 1 или два)

И все прочие выборки делаем с учетом типа.

Если из простой надо перейти в цикл или select выполнить, то смотрим в том числе на тип.

**Reply by Ekhorikov2** [reply]
Всё получилось, спасибо большое за ответы и мнения! 
У меня только тут вопрос возник по нагрузке - я в своём случае, для определения типа документа, использовал не 1 и 2 а именно фразы "Проект" и "Заказ поставщика"! Для чего? Так проще формировать наименования документов в полях. 
=: str`#tip_dokumenta ++ #nomer_dokumenta` 
Результат= "Проект 15"
Только вот правильно ли это, не влечёт чрезмерную нагрузку за собой такое решение? Или я преувеличиваю?  С числовым значением анализируется максимум две цифры, а тут, несколько букв. Или фигня всё это?

**Reply by ttm-alexeypolunin** [reply]
У меня нервный тик на использование Кириллицы, но он из древних времен кодировок CP1251 — в UTF 8 в которой работает тотум все будет норм. Сравнение тяжелее, но это экономия на спичках. Однако без лишней необходимости лучше так не делать :)

**Reply by Ekhorikov2** [reply]
Да я так и понял, но вот лень матушка, как-то так проще показалось, отмазался от возни с отображениями ))) Хотя я щас понял что можно было ещё проще сделать - передавать значение 1 и 2 и через if подставлять если 1 то =: str "Проект" ++ #nomer_dokumenta
Если 2 то =: str "Заказ поставщику" ++ #nomer_dokumenta

**Reply by ttm-alexeypolunin** [reply]
Отображение кстати можно получить кодом `#s.field_name` — это обращение вызывает расчет Селекта. Но для простых случаев или разовых запросов отлично подходит.

**Comment by mihail-cornev** [comment]
Подскажите, а как вы сделали номер вида 00019? Не понимаю как это можно реализовать, чтобы отображались нули в начале числа.

**Reply by Ekhorikov2** [reply]
> Подскажите, а как вы сделали номер вида 00019? Не понимаю как это можно реализовать, чтобы отображались нули в начале числа.

Добрый день.
Берём поле, в котором рассчитывается номер (в строке простой таблицы или в расчётной цикла). Используем функцию strAdd, в разделе Код, чтоб склеить нужные нам значения (ноли и idшник)
= : strAdd( str: $null; str: #id)
повторяем ноли за минусом самого idшника, чтоб при десятках сотнях и тысячах убирался ноль
null: strRepeat(str: "0"; num: $lenCalc)
lenCalc: 5 - $len
Берём сам idшник
len: strLength(str: #id)
или  len: strLength(str: $#nci) если это цикл.

В коде это выглядит следующим образом:
![image](https://user-images.githubusercontent.com/112487577/202499509-574d059e-fda0-4969-a577-9614d3ff1f3f.png)
